src/extensions/house/houseext.cpp

```diff
@@ -190,6 +215,756 @@ void HouseClassExtension::Object_CRC(CRCEngine &crc) const
+                return UnitTypes[BuildNavalUnit];
+            }
+        } else {
+            if (This()->BuildUnit != UNIT_NONE) {
+                return UnitTypes[This()->BuildUnit];
+            }
+        }
+        return nullptr;
+
+        /*
+        **  Infantry construction is based on the rule that up to twice the number required
+        **  to fill all teams will be created.
+        */
+    case RTTI_INFANTRY:
+    case RTTI_INFANTRYTYPE:
+        if (This()->BuildInfantry != INFANTRY_NONE) {
+            return InfantryTypes[This()->BuildInfantry];
+        }
+        return nullptr;
+
+        /*
+        **  Building construction is based upon the preconstruction list.
+        */
+    case RTTI_BUILDING:
+    case RTTI_BUILDINGTYPE:
+        if (This()->BuildStructure != STRUCT_NONE) {
+            return BuildingTypes[This()->BuildStructure];
+        }
+        return nullptr;
+    }
+    return techno;
+}
+
+
+/**
+ *  Reimplementation of HouseClass::AI_Unit.
+ *
+ *  @author: ZivDero
+ */
+int HouseClassExtension::AI_Unit()
+{
+    if (This()->BuildUnit != UNIT_NONE) return TICKS_PER_SECOND;
+
+    int harv = This()->ActiveUQuantity.Count_Of(This()->Get_First_Ownable(Rule->HarvesterUnit)->HeapID);
+    int ref = This()->ActiveBQuantity.Count_Of(This()->Get_First_Ownable(Rule->BuildRefinery)->HeapID);
+    int mult;
+    if (Session.Type == GAME_NORMAL || This()->Difficulty == DIFF_HARD) {
+        mult = 1;
+    } else {
+        mult = 2;
+    }
+
+    /*
+    **  A computer controlled house will try to build a replacement
+    **  harvester if possible.
+    */
+    if (This()->IQ >= Rule->IQHarvester && !This()->IsTiberiumShort && !This()->Is_Human_Player() && ref * mult > harv) {
+        if (This()->Get_First_Ownable(Rule->HarvesterUnit)->TechLevel <= This()->Control.TechLevel) {
+            This()->BuildUnit = This()->Get_First_Ownable(Rule->HarvesterUnit)->HeapID;
+            return TICKS_PER_SECOND;
+        }
+    }
+
+    int counter[1000]; // size increased replicating ts-patches
+    int value[std::size(counter)];
+    memset(counter, 0x00, sizeof(counter));
+    for (int& i : value) {
+        i = 0x7FFFFFFF;
+    }
+
+    /*
+    **  Build a list of the maximum of each type we wish to produce. This will be
+    **  twice the number required to fill all teams.
+    */
+    for (TeamClass* tptr : Teams) {
+        if (tptr != nullptr) {
+
+            int val = tptr->field_40;
+
+            if (((tptr->Class->IsReinforcable && !tptr->IsFullStrength) || (!tptr->IsForcedActive && !tptr->IsHasBeen)) && tptr->House == This()) {
+                DynamicVectorClass<TechnoTypeClass const*> _members;
+                tptr->Team_Members(_members);
+
+                for (int subindex = 0; subindex < _members.Count(); subindex++) {
+
+                    UnitTypeClass const* memtype = static_cast<UnitTypeClass const*>(_members[subindex]);
+
+                    if (memtype->RTTI == RTTI_UNITTYPE) {
```

Comment:
You're casting the team member to `UnitTypeClass` before you've checked its RTTI to confirm that it is actually `UnitTypeClass`. While this probably technically works due to how the class hierarchy is and how vtables are structured, it seems unsafe.